sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -205,25 +226,194 @@ static ssize_t gcoap_fileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size
+            ret = fd;
+            goto CLEAN_ON_ERROR;
+        }
+        if (create) {
+            if ((ret = vfs_lseek(fd, 0, SEEK_END)) < 0) {
+                vfs_close(fd);
+                goto CLEAN_ON_ERROR;
+            }
+            if (block1.offset != (unsigned)ret) {
+                /* expect block to be in the correct order during initial creation */
+                vfs_close(fd);
+                ret = COAP_CODE_REQUEST_ENTITY_INCOMPLETE;
+                goto CLEAN_ON_ERROR;
+            }
+        }
+        else {
+            if (block1.offset > (unsigned)stat.st_size) {
+                /* after initial file creation, random access is fine,
+                   as long as it does not result in holes in the file */
+                vfs_close(fd);
+                ret = COAP_CODE_REQUEST_ENTITY_INCOMPLETE;
+                goto CLEAN_ON_ERROR;
+            }
+            if ((ret = vfs_lseek(fd, block1.offset, SEEK_SET)) < 0) {
+                vfs_close(fd);
+                goto CLEAN_ON_ERROR;
+            }
+        }
+        if ((ret = vfs_write(fd, pdu->payload, pdu->payload_len)) < 0 ||
+            (unsigned)ret != pdu->payload_len) {
+            vfs_close(fd);
+            goto CLEAN_ON_ERROR;
+        }
+        vfs_close(fd);
+    }
+    if (!block1.more) {
+        if ((ret = vfs_stat(request->namebuf, &stat)) < 0) {
+            goto CLEAN_ON_ERROR;
+        }
+        if (create) {
+            char path[COAPFILESERVER_PATH_MAX];
+            strcpy(path, request->namebuf);
+            ret = strlen(path) - 1;
+            while(path[ret] != '/'){
+                ret--;
+            }
+            memmove(&path[ret + 1], &path[ret + 2], strlen(&path[ret + 2]) + 1);
+            if ((ret = vfs_rename(request->namebuf, path)) < 0) {
+                goto CLEAN_ON_ERROR;
+            }
+        }
+        stat_etag(&stat, &etag); /* Etag after write */
+        gcoap_resp_init(pdu, buf, len, create ? COAP_CODE_CREATED : COAP_CODE_CHANGED);
+        coap_opt_add_opaque(pdu, COAP_OPT_ETAG, &etag, sizeof(etag));
+    }
+    else {
+        gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTINUE);
+        block1.more = true; /* resource is created atomically */
+        coap_opt_add_block1_control(pdu, &block1);
+    }
+    return coap_opt_finish(pdu, COAP_OPT_FINISH_NONE);
+
+CLEAN_ON_ERROR:
+    if (create) {
+        vfs_unlink(request->namebuf);
+    }
+    return gcoap_fileserver_error_handler(pdu, buf, len, ret);
+}
+
+static ssize_t _delete_file(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                            struct requestdata *request,
+                            gcoap_fileserver_entry_t *resource)
 {
+    (void)resource;
+    int ret;
+    uint32_t etag;
+    struct stat stat;
+    if ((ret = vfs_stat(request->namebuf, &stat)) < 0) {
+        return gcoap_fileserver_error_handler(pdu, buf, len, ret);
+    }
+    if (request->options.exists.if_match && request->options.if_match_len) {
```

Comment:
So this is to only delete the file if the etag matches, right?

Reply:
yes